File: Difficulty: Hard/Longest Prefix Suffix/longest-prefix-suffix.java
//{ Driver Code Starts
// Initial Template for Java

import java.io.*;
import java.util.*;

class GFG {
    public static void main(String args[]) throws IOException {
        BufferedReader read = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(read.readLine());
        while (t-- > 0) {

            String s = read.readLine();
            Solution ob = new Solution();
            System.out.println(ob.lps(s));
        }
    }
}
// } Driver Code Ends


// User function Template for Java

class Solution {
    int lps(String s) {
        // code here
        int length = s.length();
        int lps[] = new int[length]; // making an array equal to pattern length 
        lps[0]=0; // always zero 
        int i =1;
        int len = 0;
        while(i<length){
        if(s.charAt(i)==s.charAt(len)){
        len++;
        lps[i]=len;
        i++;

        }else{
            if(len!=0){
                len = lps[len-1];
            } else{
           lps[i]=0;
           i++;

            }
        }
        }
        return  lps[length-1];
    }
}

File: Difficulty: Hard/Longest Prefix Suffix/longest-prefix-suffix.cpp
//{ Driver Code Starts
// Initial template for C++

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends

// User function template for C++

class Solution {
  public:
    int longestPrefixSuffix(string &s) {
        // Your code goes here
        int n = s.length();
        
        vector<int>lps(n, 0);
        
        for(int i=1, len=0; i<n;){
            
            if(s[i]==s[len]){
                lps[i++] = ++len;
            }
            else if(len){
                len = lps[len-1]; 
            }
            else{
                lps[i++] = 0;
            }
        }
        return lps[n-1];
    }
};

//{ Driver Code Starts.

int main() {

    ios_base::sync_with_stdio(0);
    cin.tie(NULL);
    cout.tie(NULL);

    int t;
    cin >> t;
    while (t--) {
        string str;
        cin >> str;

        Solution ob;

        cout << ob.longestPrefixSuffix(str) << "\n";

        cout << "~"
             << "\n";
    }

    return 0;
}

// } Driver Code Ends